components/calendar/src/astronomy.rs

```diff
@@ -0,0 +1,657 @@
+            205.7, 85.9, 146.1,
+        ];
+        let z: [f64; 49] = [
+            0.9287892,
+            35999.1376958,
+            35999.4089666,
+            35998.7287385,
+            71998.20261,
+            71998.4403,
+            36000.35726,
+            71997.4812,
+            32964.4678,
+            -19.4410,
+            445267.1117,
+            45036.8840,
+            3.1008,
+            22518.4434,
+            -19.9739,
+            65928.9345,
+            9038.0293,
+            3034.7684,
+            33718.148,
+            3034.448,
+            -2280.773,
+            29929.992,
+            31556.493,
+            149.588,
+            9037.750,
+            107997.405,
+            -4444.176,
+            151.771,
+            67555.316,
+            31556.080,
+            -4561.540,
+            107996.706,
+            1221.655,
+            62894.167,
+            31437.369,
+            14578.298,
+            -31931.757,
+            34777.243,
+            1221.999,
+            62894.511,
+            -4442.039,
+            107997.909,
+            119.066,
+            16859.071,
+            -4.578,
+            26895.292,
+            -39.127,
+            12297.536,
+            90073.778,
+        ];
+        let mut lambda = 0.0;
+        for i in 0..x.len() {
+            lambda += x[i] * (y[i] + z[i] * c).to_radians().sin();
+        }
+        lambda *= 0.000005729577951308232;
+        lambda += 282.7771834 + 36000.76953744 * c;
+        (lambda + Self::abberation(c, moment) + Self::nutation(moment)).rem_euclid(360.0)
+    }
+
+    fn abberation(c: f64, moment: Moment) -> f64 {
+        0.0000974 * (177.63 + 35999.01848 * c).to_radians().cos() - 0.005575
+    }
+
+    /// Find the time of the new moon preceding a given Moment
+    /// (the last new moon before moment)
+    pub fn new_moon_before(moment: Moment) -> Moment {
+        Self::nth_new_moon(Self::num_of_new_moon_at_or_after(moment) - 1)
+    }
+
+    /// Find the time of the new moon following a given Moment
+    /// (the first new moon after moment)
+    pub fn new_moon_at_or_after(moment: Moment) -> Moment {
+        Self::nth_new_moon(Self::num_of_new_moon_at_or_after(moment))
+    }
+
+    // Function to find the number of the new moon at or after a given moment;
+    // helper function for new_moon_before and new_moon_at_or_after
+    fn num_of_new_moon_at_or_after(moment: Moment) -> i32 {
+        let t0: Moment = Self::nth_new_moon(0);
+        let phi = Self::lunar_phase(moment);
+        let n = ((moment - t0) / MEAN_SYNODIC_MONTH - phi / 360.0).round() as i32;
+        let mut result = n;
+        let mut iters = 0;
+        let max_iters = 245_000_000;
+        while iters < max_iters && Self::nth_new_moon(result) < moment {
+            iters += 1;
+            result += 1;
+        }
+        result
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn check_astronomy_functions() {
```

Comment:
suggestion: Break up this test into a test for each astronomy function. Even if some functions depend on others, it helps for readability (making it easy for the reader to reason about what's going on). It has the effect of decreasing the distance between the setup for a test and usage of that setup, which is also well correlated with better "readability" for those reasons.